Add IncrementalParseResult (#2272)
apple/swift-syntax: Release Notes/510.md

Reviewer: Could you add the release notes to `Release Notes/511.md` instead of `510.md` since this change isn’t targeting the 510 release?

Author: Thanks for pointing that out, it really should be in 511!

  - Issue: https://github.com/apple/swift-syntax/issues/2092
  - Pull Request: https://github.com/apple/swift-syntax/pull/2184

- `IncrementalParseTransition`:

>   - Description: The initializer `IncrementalParseTransition.init(previousTree:edits:lookaheadRanges:reusedNodeCallback:)` is marked as deprecated. The deprecated initializer emits a warning indicating the use of `IncrementalParseTransition.init(previousIncrementalParseResult:edits:reusedNodeCallback:)`.

Reviewer: Let’s write this a little bit more imperatively so that readers have an action to follow to resolve the deprecation.

```suggestion
  - Description: The initializer `IncrementalParseTransition.init(previousTree:edits:lookaheadRanges:reusedNodeCallback:)` is marked as deprecated. Use `IncrementalParseTransition.init(previousIncrementalParseResult:edits:reusedNodeCallback:)` instead.
```

  - Issue: https://github.com/apple/swift-syntax/issues/2267
  - Pull request: https://github.com/apple/swift-syntax/pull/2272

## API-Incompatible Changes

- Effect specifiers:
  - Description: The `unexpectedAfterThrowsSpecifier` node of the various effect specifiers has been removed.
  - Pull request: https://github.com/apple/swift-syntax/pull/2219

- `Parser.parseIncrementally(source:parseTransition:)` and `Parser.parseIncrementally(source:maximumNestingLevel:parseTransition:)`:

>   - Description: `Parser.parseIncrementally(source:parseTransition:)` and `Parser.parseIncrementally(source:maximumNestingLevel:parseTransition:)` now have new return type named `IncrementalParseResult`.

Reviewer: ```suggestion
  - Description: The default versions of `Parser.parseIncrementally` return a `IncrementalParseResult` instead of a tuple. Access to the struct should be compatible with the tuple in almost all cases unless the tuple is stored into a variable and then destructed or passed to a function that expects a tuple.
```
